src/main/java/de/tum/in/www1/artemis/service/ResultService.java

```diff
@@ -133,18 +131,22 @@ public void onResultNotifiedNew(Participation participation, Object requestBody)
      * @param participation of the given result.
      * @param result        from which to extract the test cases.
      */
-    private void extractTestCasesFromResult(Participation participation, Result result) {
-        ProgrammingExercise programmingExercise = (ProgrammingExercise) participation.getExercise();
+    private void extractTestCasesFromResult(ProgrammingExerciseParticipation participation, Result result) {
+        ProgrammingExercise programmingExercise = participation.getProgrammingExercise();
         boolean haveTestCasesChanged = testCaseService.generateTestCasesFromFeedbacks(result.getFeedbacks(), programmingExercise);
         if (haveTestCasesChanged) {
             // Notify the client about the updated testCases
-            Set<ProgrammingExerciseTestCase> testCases = testCaseService.findByExerciseId(participation.getExercise().getId());
-            messagingTemplate.convertAndSend("/topic/programming-exercise/" + participation.getExercise().getId() + "/test-cases", testCases);
+            Set<ProgrammingExerciseTestCase> testCases = testCaseService.findByExerciseId(participation.getProgrammingExercise().getId());
+            messagingTemplate.convertAndSend("/topic/programming-exercise/" + participation.getProgrammingExercise().getId() + "/test-cases", testCases);
         }
     }
 
-    private void notifyUser(Participation participation, Result result) {
+    @Transactional(readOnly = true)
+    public void notifyUser(ProgrammingExerciseParticipation participation, Result result) {
         if (result != null) {
+            // Avoid circular serialization issues.
+            // TODO: This results in removing the reference from the participation to the exercise.
+            // result.getParticipation().setExercise(null);
```

Comment:
Here is the issue that `@JsonIgnoreProperties` does not work for the Discriminator classes.
Therefore we have to set the exercise to null.

However as this method is called in a Transactional Context, the reference will be overriden.

Reply:
We also have this issue in other Resources that are not in a Transactional Context, e.g. ParticipationResource -> initParticipation.

Reply:
See above note about `@JsonIgnore`.